migrations/1609942742732-fixStorageProviderInconsistencies.js

```diff
@@ -0,0 +1,166 @@
+const { schoolSchema } = require('../src/services/school/model');
+
+const StorageProviderModel = mongoose.model('storageprovider65373489214', storageProviderSchema, 'storageproviders');
+const SchoolModel = mongoose.model('schools34838583553', schoolSchema, 'schools');
+
+// How to use more than one schema per collection on mongodb
+// https://stackoverflow.com/questions/14453864/use-more-than-one-schema-per-collection-on-mongodb
+
+const EMPTY_PROVIDER = 'NOT_FOUND';
+
+const getConfig = (provider) => {
+	const awsConfig = new aws.Config({
+		signatureVersion: 'v4',
+		s3ForcePathStyle: true,
+		sslEnabled: true,
+		accessKeyId: provider.accessKeyId,
+		secretAccessKey: provider.secretAccessKey,
+		region: provider.region,
+		endpointUrl: provider.endpointUrl,
+	});
+	awsConfig.endpoint = new aws.Endpoint(provider.endpointUrl);
+	return awsConfig;
+};
+
+const getAWSS3 = async (storageProvider) => {
+	const S3_KEY = Configuration.get('S3_KEY');
+
+	storageProvider.secretAccessKey = CryptoJS.AES.decrypt(storageProvider.secretAccessKey, S3_KEY).toString(
+		CryptoJS.enc.Utf8
+	);
+	return new aws.S3(getConfig(storageProvider));
+};
+
+const getBuckets = async (provider) => {
+	const s3 = await getAWSS3(provider);
+	const response = await s3.listBuckets().promise();
+	return response.Buckets ? response.Buckets.map((b) => b.Name) : [];
+};
+
+const getSchoolsWithoutBucketsForProvider = (providerSchools, buckets) => {
+	const schoolsWithThisProvider = providerSchools[0].schools;
+	const schoolIds = schoolsWithThisProvider.map((s) => s._id);
+	const schoolsWithoutBuckets = [];
+	for (const schoolId of schoolIds) {
+		const bucketExists = buckets.indexOf(`bucket-${schoolId.toString()}`) >= 0;
+		if (!bucketExists) {
+			schoolsWithoutBuckets.push(schoolId);
+		}
+	}
+	return schoolsWithoutBuckets;
+};
+
+const getBucketsPerProvider = async (storageProviders) => {
+	const bucketsForProvider = {};
+	for (const provider of storageProviders) {
+		// eslint-disable-next-line no-await-in-loop
+		bucketsForProvider[provider._id] = await getBuckets(provider);
+	}
+	return bucketsForProvider;
+};
+
+const getProviderForSchool = (bucketsForProvider, school) => {
+	for (const [provider, buckets] of Object.entries(bucketsForProvider)) {
+		const bucketExists = buckets.indexOf(`bucket-${school.toString()}`) >= 0;
+		if (bucketExists) return provider;
+	}
+	return EMPTY_PROVIDER;
+};
+
+// eslint-disable-next-line no-unused-vars
+const printSchoolsForProvider = (foundSchoolsPerProvider) => {
+	for (const [provider, schools] of Object.entries(foundSchoolsPerProvider)) {
+		info(`For provider ${provider} found schools: ${JSON.stringify(schools)}`);
+	}
+};
+
+const updateProvidersForSchools = async (foundSchoolsPerProvider) => {
+	for (const [provider, schools] of Object.entries(foundSchoolsPerProvider)) {
+		if (provider !== EMPTY_PROVIDER) {
+			// eslint-disable-next-line no-await-in-loop
+			const result = await SchoolModel.updateMany(
```

Comment:
i must recommend taking a logging safe approach. can't we first fetch all schools, then update them each by each and log the before and after state/provider? however to see the results, we must use error log :/